From: Alex Gonzalez <alexg@balena.io>
Date: Mon, 28 Sep 2020 18:30:48 +0200
Subject: [PATCH 1/3] kexecboot: Add configuration option to boot directly the
 first entry

The --enable-default-boot configuration parameter avoids displaying a
menu and boot directly into the first entry.

Signed-off-by: Alex Gonzalez <alexg@balena.io>
---
 configure.ac    | 6 ++++++
 src/kexecboot.c | 6 ++++++
 2 files changed, 12 insertions(+)

diff --git a/configure.ac b/configure.ac
index 81c76981bcf5..df6c0b79a9cb 100644
--- a/configure.ac
+++ b/configure.ac
@@ -76,6 +76,7 @@ AC_ARG_ENABLE([no-dtb],[AS_HELP_STRING([--enable-no-dtb],[kexec after v. 2.0.13
 AC_ARG_ENABLE([no-checks],[AS_HELP_STRING([--enable-no-checks],[kexec fast reboot, no memory integrity checks @<:@default=no@:>@])],[],[enable_no_checks=no])
 AC_ARG_ENABLE([kexec-file-syscall],[AS_HELP_STRING([--enable-kexec-file-syscall],[Use the new file based syscall for kexec operation @<:@default=no@:>@])],[],[enable_kexec_file_syscall=no])
 AC_ARG_ENABLE([kexec-syscall],[AS_HELP_STRING([--enable-kexec-syscall],[Use the old kexec_load syscall for compatibility @<:@default=no@:>@])],[],[enable_kexec_syscall=no])
+AC_ARG_ENABLE([default-boot],[AS_HELP_STRING([--enable-default-boot],[Launch first menu item without displaying menu @<:@default=no@:>@])],[],[enable_default_boot=no])
 
 # args for ubiattach
 AC_ARG_WITH([ubiattach-binary],[AS_HELP_STRING([--with-ubiattach-binary="path"],[look for ubiattach binary at path @<:@default="/usr/sbin/ubiattach"@:>@])],[
@@ -264,6 +265,11 @@ AS_IF([test "x$enable_kexec_syscall" != "xno"],
 		AC_DEFINE([USE_KEXEC_SYSCALL], [1], [Define if you want to pass -c, --kexec-syscall])
 		], [])
 
+AS_IF([test "x$enable_default_boot" != "xno"],
+		[
+		AC_DEFINE([USE_DEFAULT_BOOT], [1], [Define if you want to pass  --default-boot])
+		], [])
+
 # tests for ubiattach args
 AS_IF([test "x$with_ubiattach_binary" != "xno"],
 		[
diff --git a/src/kexecboot.c b/src/kexecboot.c
index cf49b40fa480..40506b2756cf 100644
--- a/src/kexecboot.c
+++ b/src/kexecboot.c
@@ -1091,6 +1091,9 @@ int main(int argc, char **argv)
 		exit(-1);
 	}
 
+#ifdef USE_DEFAULT_BOOT
+	rc = A_DEVICES;
+#else
 	/* Collect input devices */
 	inputs_init(&inputs, 8);
 	inputs_open(&inputs);
@@ -1099,6 +1102,7 @@ int main(int argc, char **argv)
 	/* Run main event loop
 	 * Return values: <0 - error, >=0 - selected item id */
 	rc = do_main_loop(&params, &inputs);
+#endif
 
 #ifdef USE_FBMENU
 	if (params.gui) {
@@ -1112,8 +1116,10 @@ int main(int argc, char **argv)
 		if (ttyfp != stdout) fclose(ttyfp);
 	}
 #endif
+#ifndef USE_DEFAULT_BOOT
 	inputs_close(&inputs);
 	inputs_clean(&inputs);
+#endif
 
 	log_close(lg);
 	lg = NULL;
